cmd/otel-allocator/target/discovery.go

```diff
@@ -105,43 +127,122 @@ func (m *Discoverer) ApplyConfig(source allocatorWatcher.EventSource, scrapeConf
 	return m.manager.ApplyConfig(discoveryCfg)
 }
 
-func (m *Discoverer) Watch(fn func(targets map[string]*Item)) error {
-	labelsBuilder := labels.NewBuilder(labels.EmptyLabels())
+func (m *Discoverer) Run() error {
+	err := m.run(m.manager.SyncCh())
+	if err != nil {
+		m.log.Error(err, "Service Discovery watch event failed")
+		return err
+	}
+	<-m.close
+	m.log.Info("Service Discovery watch event stopped: discovery manager closed")
+	return nil
+}
+
+// UpdateTsets updates the target sets to be scraped.
+func (m *Discoverer) UpdateTsets(tsets map[string][]*targetgroup.Group) {
+	m.mtxScrape.Lock()
+	m.targetSets = tsets
+	m.mtxScrape.Unlock()
+}
+
+// reloader triggers a reload of the scrape configs at regular intervals.
+// The time between reloads is defined by reloadIntervalDuration to avoid overloading the system
+// with too many reloads, because some service discovery mechanisms can be quite chatty.
+func (m *Discoverer) reloader() {
+	reloadIntervalDuration := model.Duration(5 * time.Second)
+	ticker := time.NewTicker(time.Duration(reloadIntervalDuration))
+
+	defer ticker.Stop()
+
 	for {
 		select {
 		case <-m.close:
-			m.log.Info("Service Discovery watch event stopped: discovery manager closed")
-			return nil
-		case tsets := <-m.manager.SyncCh():
-			m.ProcessTargets(labelsBuilder, tsets, fn)
+			return
+		case <-ticker.C:
+			select {
+			case <-m.triggerReload:
+				m.Reload()
+			case <-m.close:
+				return
+			}
 		}
 	}
 }
 
-func (m *Discoverer) ProcessTargets(builder *labels.Builder, tsets map[string][]*targetgroup.Group, fn func(targets map[string]*Item)) {
+// Reload triggers a reload of the scrape configs.
+// This will process the target groups and update the targets concurrently.
+func (m *Discoverer) Reload() {
+	m.mtxScrape.Lock()
+	var wg sync.WaitGroup
 	targets := map[string]*Item{}
+	timer := prometheus.NewTimer(processTargetsDuration)
+	defer timer.ObserveDuration()
+
+	for jobName, groups := range m.targetSets {
+		wg.Add(1)
+		// Run the sync in parallel as these take a while and at high load can't catch up.
+		go func(jobName string, groups []*targetgroup.Group) {
+			processedTargets := m.processTargetGroups(jobName, groups)
+			m.mtxTargets.Lock()
+			for k, v := range processedTargets {
+				targets[k] = v
+			}
+			m.mtxTargets.Unlock()
+			wg.Done()
+		}(jobName, groups)
+	}
+	m.mtxScrape.Unlock()
```

Comment:
I think this doesn't run into any race conditions where `m.targetSets` is changed while the goroutines run, but it'd be better to make it more explicit imo:

```suggestion
	targetSets := m.targetSets // we never modify this map, just replace it, so this is thread-safe
	var wg sync.WaitGroup
	targets := map[string]*Item{}
	timer := prometheus.NewTimer(processTargetsDuration)
	defer timer.ObserveDuration()

	for jobName, groups := range targetSets {
		wg.Add(1)
		// Run the sync in parallel as these take a while and at high load can't catch up.
		go func(jobName string, groups []*targetgroup.Group) {
			processedTargets := m.processTargetGroups(jobName, groups)
			m.mtxTargets.Lock()
			for k, v := range processedTargets {
				targets[k] = v
			}
			m.mtxTargets.Unlock()
			wg.Done()
		}(jobName, groups)
	}
```

Reply:
I'm mostly fine with that change, I'm just wondering why move to a variable @swiatekm ?

Reply:
I removed the `mtxScrape` lock, so `m.targetSets` can be changed while we're processing. In your previous code, `range m.targetSets` is only evaluated once, at the start of the loop, so the effect is very similar. I created a new variable instead to make this clearer.